File: calculator.lua
function Calculator() -- doesn't need a function
  print("***************************")
  print("A SIMPLE CALCULATOR IN LUA")
  print("ENTER INPUTS WITHOUT SPACES")
  print("***************************\n\n")

  while true do
    io.write("Enter an operation (+, -, *, /) | 'q' to quit: ")
    local option = io.read()
    if option == 'q' or option == 'Q' then
      break
    end
    io.write("Enter number 1 (double): ")
    local num_1 = tonumber(io.read())
    io.write("Enter number 2 (double): ")
    local num_2 = tonumber(io.read())
    local total

    -- Lua has no switch cases and I don't feel like implementing my own
    if option == '+' then
      total = num_1 + num_2
    elseif option == '-' then
      total = num_1 - num_2
    elseif option == '*' then
      total = num_1 * num_2
    else
      total = num_1 / num_2 -- Default is division | could've just dont if instad of else-if statements
    end
    print("The result is: " .. total)
  end
end

Calculator() -- call Calculator function

File: calculator.zig
const std = @import("std");

pub fn main() !void { // can return an error
        var stdout_buffer: [1024]u8 = undefined;
        var stdout_writer = std.fs.File.stdout().writer(&stdout_buffer);
        const stdout = &stdout_writer.interface;

        var stdin_buffer: [1024]u8 = undefined;
        var stdin_reader = std.fs.File.stdin().reader(&stdin_buffer);
        const stdin = &stdin_reader.interface;

	try stdout.print("***************************\n", .{}); // always expects formatted strings
	try stdout.print("A SIMPLE CALCULATOR IN ZIG\n", .{});
	try stdout.writeAll("***************************\n"); // writeAll doesn't require formatting
        try stdout.flush(); // force sends everything in the current buffer to the terminal

        while (true) {
            try stdout.writeAll("Enter an option (+, -, *, /) | 'q' to quit: ");
            try stdout.flush();

            const option_input = stdin.takeDelimiter('\n') catch break; // get length + crash on error
            if (option_input == null) break;
            const option_line = option_input.?;
            if (option_line.len == 0) continue;
            const option = option_line[0];

            if (option == 'q' or option == 'Q') break;

            const num1 = getNumber(stdout, stdin, "Enter number 1 (double): ") orelse continue; // continue on error
            const num2 = getNumber(stdout, stdin, "Enter number 2 (double): ") orelse continue;

            const total: ?f64 = switch (option) {
                '+' => num1 + num2,
                '-' => num1 - num2,
                '*' => num1 * num2,
                '/' => num1 / num2,
                else => null,
            };

            if (total) |t| { // capture to extract inner value of total since it's an optional float
                try stdout.print("The result is {d}\n", .{t});
            }
            try stdout.flush();
        }
}

fn getNumber(stdout: *std.Io.Writer, stdin: *std.Io.Reader, message: []const u8) ?f64 {
    stdout.writeAll(message) catch return null;
    stdout.flush() catch return null;
    const line_opt = stdin.takeDelimiter('\n') catch return null;
    if (line_opt == null) return null;
    return std.fmt.parseFloat(f64, line_opt.?) catch null;
}
